src/lars/backup.23

```diff
@@ -0,0 +1,266 @@
+TERMIN
+TERMIN
+]
+
+A==1
+B==A+1
+C==3
+D==4
+P==17
+
+CLOC==1		;Core link channel for DUMP output.
+CLIC==2		;Core link channel for input to DUMP.
+LOGC==3		;Log channel.  DUMP output is written here.
+DSKC==4		;Disk file channel for TS DUMP.
+USRC==5		;Inferior job channel.
+
+PDLLEN=100
+
+DEFINE SYSCAL A,B
+	.CALL [SETZ ? SIXBIT/A/ ? B ((SETZ))]
+TERMIN
+
+DEFINE LOOK [ZZZ
+	MOVEI A,ZZZ
+	PUSHJ P,LOOK1
+TERMIN
+
+DEFINE SEND [ZZZ
+	MOVEI A,ZZZ
+	PUSHJ P,SEND1
+TERMIN
+
+LOC 42
+	TSINT
+
+LOC 100
+GO:	MOVE P,[-PDLLEN,,PDL]
+
+	;Open two core links for DUMP TTY: input/output.
+	SYSCAL OPEN,[[.UAI,,CLOC] ? CLODEV ? CLOFN1 ? CLOFN2 ? CLODIR]
+	 .LOSE %LSFIL
+	SYSCAL OPEN,[[.UAO,,CLIC] ? CLIDEV ? CLIFN1 ? CLIFN2 ? CLIDIR]
+	 .LOSE %LSFIL
+
+	;Delete the old log file.  Then open it for writing.
+	SYSCAL DELETE,[LOGDEV ? LOGFN1 ? LOGFN2 ? LOGDIR]
+	 JFCL
+	SYSCAL OPEN,[[.UAO,,LOGC] ? LOGDEV ? LOGFN1 ? LOGFN2 ? LOGDIR]
+	 .LOSE %LSFIL
+
+	;Make an inferior job.
+	.OPEN USRC,[.BIO,,'USR ? 0 ? SIXBIT /DUMP/]
+	 .LOSE %LSSYS
+	.RESET USRC,
+	.USET USRC,[.RINTB,,INFBIT']	;Inferior interrupt bit.
+	.SUSET [.SMSK2,,INFBIT]		;Subscribe to those interrupts.
+
+	;Translate inferior TTY: input and output to the core links.
+	SYSCAL TRANAD,[[1,,USRC] ? TTYTRA ? CLITRA]
+	 .LOSE %LSSYS
+	SYSCAL TRANAD,[[2,,USRC] ? TTYTRA ? CLOTRA]
+	 .LOSE %LSSYS
+
+	;Load inferior from TS DUMP.
+	SYSCAL OPEN,[[.UII,,DSKC] ? DMPDEV ? DMPFN1 ? DMPFN2 ? DMPDIR]
+	 .LOSE %LSFIL
+	SYSCAL LOAD,[MOVEI USRC ? MOVEI DSKC]
+	 .LOSE %LSSYS
+	.IOT DSKC,A			;Get start address.
+	HRRZS A
+	.USET USRC,[.SUPC,,A]
+	.USET USRC,[.SUSTP,,[0]]	;Start inferior.
+	.CLOSE DSKC,
+
+	;Convert today's date to an ASCII string.
+	.RDATE A,
+	MOVE C,[440700,,DATE]
+LOOP2:	LDB D,[360600,,A]
+	ADDI D,40
+	IDPB D,C
+	LSH A,6
+	JUMPN A,LOOP2
+	IDPB A,C
+
+	PUSHJ P,TAPES			;Figure out highest tape number.
+	AOS TAPE			;Add one to it.
+	PUSHJ P,MOUNT			;Mount remote tape.
+	PUSHJ P,DUMPI			;Incremental backup.
+	PUSHJ P,QUIT
+
+HANG:	JFCL				;Wait for DUMP to finish.
+	 .HANG
+	JRST HANG
+
+TAPES:	SETZM TAPE
+	LOOK PROMPT			;Send the TAPES command,
+	SEND [ASCIZ /TAPES�/]
+	LOOK [ASCIZ /LIST DEV =/]
+	SEND CR
+TAPES1:	PUSHJ P,INPUT
+	CAIE D,^J			;Look for a new line.
+	 JRST TAPES1
+	MOVEI A,6
+	MOVEI B,0
+TAPES2:	PUSHJ P,INPUT			;Parse a decimal number.
+	CAIN D,^M
+	 POPJ P,			;Empty line means end of list.
+	CAIN D,40
+	 SKIPA D,[0]
+	  SUBI D,"0
+	IMULI B,10.
+	ADD B,D
+	SOJG A,TAPES2
+	CAMLE B,TAPE			;Save the largest tape number.
+	 MOVEM B,TAPE
+	JRST TAPES1
+
+MOUNT:	LOOK PROMPT			;Send the REMOTE command,
+	SEND [ASCIZ /REMOTE�/]		;and answer the questions.
+	LOOK [ASCIZ /HOST=/]
+	SEND HOST
+	SEND CR
+	LOOK [ASCIZ /DRIVE=/]
+	PUSHJ P,DRIVE
+	LOOK [ASCIZ /READ-ONLY? /]
+	SEND [ASCIZ /N/]
+	POPJ P,
+
+DUMPI:	LOOK PROMPT			;Send DUMP I command.
+	SEND [ASCIZ /DUMP I�/]
+	LOOK [ASCIZ /TAPE NO=/]
+	PUSHJ P,TAPENO
+	SEND CR
+	LOOK [ASCIZ /E-O-T/]
+	POPJ P,
+
+DUMPF:	LOOK PROMPT			;Send DUMP F command.
```

Comment:
How does one specify to do a full, versus incremental dump? I don’t see DUMPF is ever reached.

Reply:
There is no way to specify a full dump.  The code is in there for future extension, but it's not reachable now.  Maybe if the JNAME is DUMPF or something?

Reply:
Yes, that's a good idea -- use the JNAME.  Wanna update before merging, or wait for another PR?

Reply:
Turns out full dump doesn't work yet.  Let's get this in and figure out improvements later.